### ISSUE TITLE
Core functionality thoughts

### ISSUE THREAD

**Issue by t00h** [t00h/Python-Single-Character-Compression#1]
I'm wondering exactly how to implement the core functionality of the program. By the time the file's content data is passed to the loop to compress it, it's been converted to binary form - probably implementing a Unicode encoding to allow for all characters (or as many as possible) in the input file.

From there, I think at each loop I'll replace each possible pair of characters in the current set with one character in the next set. For example, from the first set (binary): 0,0 --> 2; 0,1 --> 3; 1,0 --> 4; 1,1 --> 5

The possible problem is that the number of possible pairing combinations at each set is the square of the previous (e.g. for 2, 3, 4, and 5 you can have 16 possible pairings, then 256, etc). This means that I might reach the bit limit of whichever encoding I choose fairly quickly.

I might, therefore add a conditional in the compression loop meaning that the loop before this limit is reached (where characters with number values greater than the maximum of the encoding), convert the String back to binary again, and resume the loop, but counting the number of times additional binary conversions are needed: otherwise backwards decompression won't be possible.

The process of halving the numbers a few times then converting them back to binary form might not actually reduce the number of bits needed to store the information - e.g. if using an encoding with a small number of bits, so only a couple of encodings can occur before having to convert back to binary.

For example, ASCII has 128 characters that can be encoded. That means we could only halve the number of bits twice before converting them back to binary, which would increase the number of bits by 8 (or 7 if avoiding the initial signed bit) times. Obviously, this doesn't make sense when trying to compress a file.

So I might need to avoid encodings altogether and in their place use pure numbers only, with some class akin to BigIntegers in Java (because we've established they'll get big fast), with successive sets only being numbers from the binary 0 and 1, no longer representing characters. That would let me store the information in a file as 1 (albeit very big) number. I probably need to establish, again, these numbers for cases other than very small files, would definitely reduce the number of bits required to store the information in the file.

Thinking more about a number-based system, maybe I could recursively turn the numbers at some size back into binary form, then running them again, similarly as I thought out for using characters as above - but past the limit of character encodings. I imagine there's probably some minimum limit for the number size to do this at whilst still reducing the file size, but past that there's probably a nice point to choose performance-wise: balancing doing looping operations on big sets of big numbers getting very big very fast vs recursively calling a function.

I probably need to think it out some more before implementing, regardless of my chosen approach. I also first need to implement the functions to:
- convert a String taken from the data held in a file into binary form
- generate a MD5 checksum from a file (and/or the data read in from it)
- generate a SHA checksum from a file (and/or the data read in from it)

TLDR: this project is pretty complicated to approach because I'm trying to find a way of representing a lot of data in 1 character/number (probably number), and I need to do a lot of thinking about it. Also, I need to implement some other methods first.
